### ISSUE TITLE
No output with docker when run non-interactively

### ISSUE THREAD

**Issue by jodh-intel** [01org/cc-oci-runtime#17]
If the runtime is configured for docker and `docker run` is specified without `-ti`, no output is currently generated to either the console or the docker log (`docker logs` command).

```
$ docker-upstream run --runtime runc ubuntu echo "hello world"
hello world

$ docker-upstream logs  $(docker-upstream ps -qa | head -1)
hello world

$ docker-upstream run --runtime cor ubuntu echo "hello world"
container 20ed5890250df32cd7c1b24027bcba708c6b350481557205fc881a16e924b27d created

$ docker-upstream logs  $(docker-upstream ps -qa | head -1)
container 20ed5890250df32cd7c1b24027bcba708c6b350481557205fc881a16e924b27d created

$ docker-upstream run -it --runtime runc ubuntu echo "hello world"
hello world

$ docker-upstream logs  $(docker-upstream ps -qa | head -1)
hello world

$ docker-upstream run -it --runtime cor ubuntu echo "hello world"
hello world
```

**Comment by amshinde** [comment]
Docker creates name pipes init-std* under /run/docker/libcontainerd/{container} dir. These pipes are used for reading the container output. containerd-shim creates unnamed pipes (https://github.com/docker/containerd/blob/ef23e7041b27a773da6dfaf6483f018e48c40d79/containerd-shim/process.go#L311)  to pipe the IO from from the runtime to the shim. (https://github.com/docker/containerd/blob/ef23e7041b27a773da6dfaf6483f018e48c40d79/containerd-shim/process.go#L173). The shim copies the output from the unnamed pipes to the named ones for the output to be available for docker.
Runc simply dups the  standard IO when run in non-interactive mode.  The solution for our runtime is to  pipe the output from the qemu process to stdout of the parent process.  This should solve the non-interactive output as well as fix docker attach.

**Comment by grahamwhaley** [comment]
I've not read the code, but your analysis sounds good to me.

**Comment by devimc** [comment]
@amshinde  :+1:

**Comment by amshinde** [comment]
Passed  'stdio,id=charconsole0,signal=off'  to qemu as its virtual console device for qemu to use the stdio inherited from the parent. Following suggestion from devimc, added a delay before  the parent exits. The output from qemu  is shown as long as the parent is alive. Working on a fix.

**Comment by amshinde** [comment]
I experimented with the stdio option for qemu few days back. For this option to work, looks like qemu needs the parent to be around. Doing an extra fork before forking qemu resolves this issue. I have code that forks the runtime itself before forking qemu in an experimental branch that works. 
So, the approach of using cc-shim in the future will help  tackle this issue.

**Comment by gorozco1** [comment]
I think there are two issues in this:

First part of this bug is already fixed, while the second part:

**cor running in the background when not use -it VS runc exiting immediately is another issue...**

Can we close this and keep with the issue #316 ??

**Comment by amshinde** [comment]
This was resolved using a workaround, I wanted to keep it open till we end up with a robust solution
(not writing directly to docker pipes). I am ok closing this one.

**Comment by amshinde** [comment]
Will be fixed with #271

**Comment by GabyCT** [comment]
This issue was already fixed and it is working properly
# docker run busybox echo "containers"
containers
# docker logs $(docker ps -qa | head -1)
containers

We already have tests for run non-interactively and logs.
